feat: added tls config

Avi98 opened the pull request:
### Component/Part
backend -> Database cofig 

### Description
This PR fixes #3063

### Steps


- [x] Added implementation
- [x] Added / updated tests
- [x] Added / updated documentation
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

ErikMichelson commented on docs/content/references/config/database.md:
```diff
@@ -7,12 +7,48 @@ We officially support and test these databases:
+
+| environment variable                  | default | type   | description                                                                                                   |
+| ------------------------------------- | ------- | ------ | ------------------------------------------------------------------------------------------------------------- |
+| `HD_DATABASE_SSL_CA`                  | -       | string | The SSL/TLS certificate key needed for database connection.                                                   |
```
Maybe specify more concrete: Do you pass the file name of the CA certificate or do you pass the contents (and if so in which encoding)?

Please state that this certificate (as every CA cert) is the anchor for checking the validity of the cert chain.

ErikMichelson commented on docs/content/references/config/database.md:
```diff
@@ -7,12 +7,48 @@ We officially support and test these databases:
+| `HD_DATABASE_SSL_CA`                  | -       | string | The SSL/TLS certificate key needed for database connection.                                                   |
+| `HD_DATABASE_SSL_REJECT_UNAUTHORIZED` | 'false' | string | The server certificate is verified against the list of supplied CAs                                           |
+| `HD_DATABASE_SSL_KEY`                 | -       | string | This is passed as the key option to [SSL key option](https://node-postgres.com/features/ssl#self-signed-cert) |
+| `HD_DATABASE_SSL_CERT`                | -       | string | The client certificate to use in the SSL handshake                                                            |
```
Same as for the CA certificate.

ErikMichelson reviewed (dismissed):
Thanks for your contribution!

I haven't tested the code yet, these are just some things I saw when having a look at the code.

Avi98 commented:
> Thanks for your contribution!
> 
> I haven't tested the code yet, these are just some things I saw when looking at the code.

Erik, I have added an env flag `HD_DATABASE_SSL` for validating the SSL/TLS config. If the flag is absent, the SSL/TLS joi validation will not work. 


| environment variable | default | example          | description                                                                                |
| -------------------- | ------- | ---------------- | ------------------------------------------------------------------------------------------ |
| `HD_DATABASE_SSL`    | '0'     | `1`              | Pass this value when SSL/TLS configuration is needed                                      |

Tested joi validation for following condition
1. When `HD_DATABASE_SSL=1`  & `HD_DATABASE_TYPE=postgres` SSL/TLS env config for postgres should be present.  
2. When `HD_DATABASE_SSL=1`  & `HD_DATABASE_TYPE=mysql` SSL/TLS env config SSL/TLS for mysql/Mariadb should be present. 
3. `HD_DATABASE_SSL` is not there (backward compatible) SSL/TLS validation is not added 

https://github.com/hedgedoc/hedgedoc/assets/26133749/088843ea-355b-4054-833f-4c5992518668

ErikMichelson commented on backend/src/config/database.config.ts:
```diff
@@ -46,8 +78,56 @@ const databaseSchema = Joi.object({
+  }),
 });
 
+const getTlsConfig = (dbType: DatabaseType, needsSSL: boolean) => {
+  if (!needsSSL) return;
+
+  const postgresTlsConfig = dbType === DatabaseType.POSTGRES && {
+    ssl: {
+      ca: getSecret(process.env.HD_PG_SSL_CA_PATH),
+      rejectUnauthorized: process.env.HD_PG_SSL_REJECT_UNAUTHORIZED === 'true',
+      key: getSecret(process.env.HD_PG_SSL_KEY_PATH),
+      cert: getSecret(process.env.HD_PG_SSL_CERT_PATH),
+    },
+  };
+
+  const sqlTlsConfig = [DatabaseType.MARIADB, DatabaseType.MYSQL].includes(
+    dbType,
+  ) && {
+    ssl: {
+      ca: getSecret(process.env.HD_SQL_SSL_CA_PATH),
+      key: getSecret(process.env.HD_SQL_SSL_KEY_PATH),
+      cert: getSecret(process.env.HD_SQL_SSL_CERT_PATH),
+      rejectUnauthorized: process.env.HD_SQL_SSL_REJECT_UNAUTHORIZED === 'true',
+      ciphers: process.env.HD_SQL_SSL_CIPHERS,
+      maxVersion: process.env.HD_SQL_SSL_MAX_VERSION,
+      minVersion: process.env.HD_SQL_SSL_MIN_VERSION,
+      passphrase: process.env.HD_SQL_SSL_PASSPHRASE,
+    },
+  };
+
+  return sqlTlsConfig || postgresTlsConfig;
+};
```
As this method is rather crucial to database security options, it should be unit-tested. The `getSecret` method could be mocked for that.

ErikMichelson reviewed (commented):
This gets in the right direction! :)

What still surprises me is the reason we need completely separate environment variables for postgres and maraidb/mysql? I get that you intended to make explicit which options are only relevant for the chosen database type. But having for example both `HD_SQL_SSL_CA_PATH` and `HD_PG_SSL_CA_PATH` seems a bit unnecessary to me. For changing that the code doesn't need greater changes, it's more a thing of renaming the variables and updating the documentation.

sndrsmnk commented:
I was looking for (PostgreSQL) TLS/SSL support in Hedgedoc when i found this. Any update on this progressing?
Thanks for the work!!

ErikMichelson committed:
feat(config): TLS config options for database connection

This was originally contributed by @Avi98 back when the config still used
Joi and TypeORM instead of zod and knex. This commit adapts the same changes
previously done but ports them over to zod and knex. Furthermore, the tests
are updated to ensure all aspects of the config are tested.

Co-authored-by: Avinash <avinash.kumar.cs92@gmail.com>
Co-authored-by: Philip Molares <philip.molares@udo.edu>
Signed-off-by: Erik Michelson <github@erik.michelson.eu>
